flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/operators/join/temporal/TemporalRowTimeJoinOperator.java

```diff
@@ -294,6 +320,13 @@ private int indexOfFirstElementNewerThanTimer(long timerTimestamp, List<RowData>
 	/**
 	 * Binary search {@code rightRowsSorted} to find the latest right row to join with {@code leftTime}.
 	 * Latest means a right row with largest time that is still smaller or equal to {@code leftTime}.
+	 * For example with: rightState = [1(+I), 4(-U), 4(+U), 7(-U), 7(+U), 9(-D), 12(I)],
```

Comment:
I think we can't store both `4(-U)` and `4(+U)` in state, because they have the same map key `4`. 

An improvement here is we don't store the `RowData` into state for retraction, but just store `null`. This can be better for performance if the `RowData` is large. We can use `TreeMap<Long, RowData>` or `List<Tuple2<Long, RowData>` to replace the current `List<RowData>`.

Reply:
I did not get the improvement point, could you explain more ? other comments has addressed

Reply:
Got your point, maybe this performance improvement is minor because (1) it only saves retraction records storage not all records (2) many retraction records between two watermarks  is not common, So I tend to improve it later if it's very important for performance